### ISSUE TITLE
[Accessibility] accordion accessibility

### ISSUE THREAD

**Issue by yzen** [devtools-html/debugger.html#4080]
This issue tracks the work to make the accordion accessible. A PR can do one or several items in the checklist:

- [x] Visually (CSS focus styling) there is no way to focus on the accordion headers and reload button
- [x] Need to be able to navigate to headers with keyboard
- [ ] Need to be able to collapse expand headers with keyboard
- [x] Headers/Panels should at least have list/listitem semantics including expanded and collapsed states.
- [ ] Headers listitem should have expanded and collapsed states exposed.
- [ ] When reaching a tree widget inside an accordion section focus should remain on the container for efficiency. The tree should handle keyboard using arrow keys (similar to source list tree widget).
- [ ] Breakpoint related check boxes must be properly accessible. Text label (or code line that serves as one) must be set as the formal label for the checkbox.

### Screenshot 
![Screen Shot 2017-09-18 at 8.12.38 PM.png](https://shipusercontent.com/623f4c6a99b4a6184f5f7815090626d6/Screen%20Shot%202017-09-18%20at%208.12.38%20PM.png)

**Comment by jasonLaster** [comment]
#### Style the focus state:

This is a patch for styling the focus state. I'm not sure what style we should apply. 

```diff
diff --git a/src/components/shared/Accordion.css b/src/components/shared/Accordion.css
index 0cd96de..3035a11 100644
--- a/src/components/shared/Accordion.css
+++ b/src/components/shared/Accordion.css
@@ -30,6 +30,10 @@
   cursor: default;
 }

+.accordion ._header:focus {
+  background-color: yellow; /* not sure what color makes sense */
+}
+
 .accordion ._header:hover {
   background-color: var(--theme-toolbar-background-hover);
 }
```


#### Need to be able to navigate to headers with keyboard

Do we need to do something to make the header focusable?

#### Need to be able to collapse expand headers with keyboard

```diff
diff --git a/src/components/shared/Accordion.js b/src/components/shared/Accordion.js
index 9f45ce2..fb69473 100644
--- a/src/components/shared/Accordion.js
+++ b/src/components/shared/Accordion.js
@@ -35,6 +35,7 @@ class Accordion extends Component {

     const self: any = this;
     self.renderContainer = this.renderContainer.bind(this);
+    self.onKeyDown = this.onKeyDown.bind(this);
   }

   componentWillReceiveProps(nextProps: Props) {
@@ -93,9 +94,15 @@ class Accordion extends Component {
     );
   }

+  // our goal is to toggle the header with spacebar and enter
+  // i'm not sure if this event listener will work when the focus is on the header
+  onKeyDown(e) {
+    // will this recieve the event?
+  }
+
   render() {
     return (
-      <div className="accordion">
+      <div className="accordion" onKeyDown={this.onKeyDown}>
         {this.props.items.map(this.renderContainer)}
       </div>
     );
```

**Comment by yzen** [comment]
### Need to be able to navigate to headers with keyboard
Header elements should be tabbable (have tabindex="0" attribute). This way, within the accordion, tabbing will jump between headers, optional controls inside headers (checkboxes, etc) and tree elements within the accordion list item (they will be keyboard accessible when the tree component is updated).

**Comment by yzen** [comment]
> what do you mean by container?
By container I mean the element that has a ```.tree``` class (e.g. element containing the tree component). This will be solved with tree component update.

**Comment by yzen** [comment]
> Headers/Panels should at least have list/listitem semantics?
> Is this defined somewhere?

With some list specific styling taken care of (removed) the following changes will make accordion more semantically reach:

```diff
--- a/src/components/shared/Accordion.js
+++ b/src/components/shared/Accordion.js
@@ -73,14 +73,14 @@ class Accordion extends Component {
       .replace(/\s/g, "-")}-pane`;
 
     return (
-      <div className={containerClassName} key={i}>
-        <div className="_header" onClick={() => this.handleHeaderClick(i)}>
+      <li className={containerClassName} key={i}>
+        <h2 className="_header" onClick={() => this.handleHeaderClick(i)}>
           <Svg name="arrow" className={opened[i] ? "expanded" : ""} />
           {item.header}
           {item.buttons ? (
             <div className="header-buttons">{item.buttons}</div>
           ) : null}
-        </div>
+        </h2>
         {created[i] || opened[i] ? (
           <div
             className="_content"
@@ -89,15 +89,15 @@ class Accordion extends Component {
             {createElement(item.component, item.componentProps || {})}
           </div>
         ) : null}
-      </div>
+      </li>
     );
   }
 
   render() {
     return (
-      <div className="accordion">
+      <ul className="accordion">
         {this.props.items.map(this.renderContainer)}
-      </div>
+      </ul>
     );
   }
 }
```

Note <H2> is used here only because I assumed it's the logical nested header level, it might actually be something else.

**Comment by CrisOrellana** [comment]
Working on bug " Need to be able to collapse expand headers with keyboard" 
.claim on that specific bug. Did not want to "/claim" for the whole thing. CS 490 
WNE

**Comment by jasonLaster** [comment]
/claim

Hey @CrisOrellana, I'll claim it so it's yours, but you can start with the easy stuff and we'll go from there. Small PRs are always better.

**Comment by CrisOrellana** [comment]
We have learned a lot while working on this enhancement as well as how long it takes just to complete something that doesn't even seem like a lot of work when it actually is. Although we have not completed the enhancement on time due to the semester ending, we figured out some things that could very well help someone who does have the time to do so! One thing we have figured out is how the accordion works in relation to accessibility. We figured out that an event listener may be needed in order to assign a specific hotkey to collapse and expand the headers along with messing with the accordion. If there are any questions please feel free to message me! 

/unclaim

**Comment by emmet0r** [comment]
/claim

I'll have a group of students taking up this issue this week! Will tag them shortly.

**Comment by chan750** [comment]
@paesh1997, @denigrise, and I are in @emmet0r class and will be trying to tackle the second check box "Need to be able to navigate to headers with keyboard"

**Comment by moses-gbondo** [comment]
@jawyoonis and @phuyalb16 are in @emmet0r class and we are working on checkbox 4 (header/panels and list semantics).

**Comment by jawyoonis** [comment]
`return (
      <li className={item.className} key={i}>
        <h2 className="_header" onClick={() => this.handleHeaderClick(i)}>
          <Svg name="arrow" className={opened ? "expanded" : ""} />
          {item.header}
          {item.buttons ? (
            <div className="header-buttons">{item.buttons}
            </div>
          ) : null}
        </h2>
        {opened && (

          <div className="_content">

            {cloneElement(item.component, item.componentProps || {})}
          </div>
        )}
        </li>
    );
  };

  render() {
    return (
      <ul className="accordion">
        {this.props.items.map(this.renderContainer)}
      </ul>`

**Comment by chan750** [comment]
@jasonLaster @yzen After inserting tabindex="0" into the <div className="_header" and the <div className="header-buttons" in the /src/components/shared/Accordion.js file, we noticed that we were able to tab through the headers in the order specified in the image attached. We could only do this after pressing Crtl Shift F to search in files, otherwise it would tab to the collapse pane buttons and then the user could no longer use tab to navigate. We do not know what happens at points 3 and 7, and it seems to stop after point 15. What order or behavior that we should be striving for? Any additional information would be helpful and appreciated.
![selection_001](https://user-images.githubusercontent.com/36204780/38105854-38f5c138-335b-11e8-9944-4c0791a24ecc.png)

**Comment by darkwing** [comment]
Hello @chan750!

Thank you for your help!  My best advice on finding out what's being focused upon between 3 and 7 is using `document.activeElement` (https://developer.mozilla.org/en-US/docs/Web/API/DocumentOrShadowRoot/activeElement) to see what element, if any, is being focused.

As for desired order, I don't know that I have a "correct" answer.  Logically, I think a "left to right" approach makes sense, i.e. the SourcesTree on the left / Outline panes, then into the document, then into the secondary pane.  @yzen Could you provide a bit more insight as for `tabIndex`?

David

**Comment by chan750** [comment]
@paesh1997 @denigrise and I were able to make some improvements to the tab order, but we will have to stop here. Our semester is coming to a close and we will no longer be able to continue working on the second check box "Need to be able to navigate to headers with keyboard." Running **setInterval(function() { console.log(document.activeElement); }, 3000);** in the web console will allow the user to see which elements are being focused. We noticed that <**div class="CodeMirror cm-s-mozilla" tabindex="0">** and **< html class="theme-light html" dir="ltr">** are being focused when they should not be. So that is where we recommend starting.

**Comment by please18** [comment]
Hello all, 
My teammates (@jawyoonis and @phuyalb16) and I were working on check box #4 -  Headers/Panels should at least have list/listitem semantics including expanded and collapsed states.
Our PR was successfully merged and therefore it is time to  /unclaim the issue. 
Thanks for your help @jasonLaster @DavideDeCenzo @yzen

**Comment by jasonLaster** [comment]
Go for it

**Comment by HeidiEllis** [comment]
+1

The WNE students are no longer working on these issues.

On 04/30/2018 02:44 PM, Jason Laster wrote:

Go for it

**Comment by yzen** [comment]
@nastark519 Feel free to ask any questions regarding the issue and possible solutions.

**Comment by emmet0r** [comment]
All yours, @nastark519 - we're done for the semester!

If I remember correctly, one student team found that one of the headers was already expandable with keyboard interaction, but that it wasn't highlighting to indicate that to the user. (They found that on accident through repeatedly pressing Tab + Space, so that might be a useful place to start.)

**Comment by nastark519** [comment]
/claim "Need to be able to collapse expand headers with keyboard"

**Comment by nastark519** [comment]
Hello @yzen do you know where or if there is an action listener for onClick to collapse expand headers? we have been looking in Accordion.js ? thank you in advance.

**Comment by yzen** [comment]
Yeah I believe you are looking in the right place - https://github.com/devtools-html/debugger.html/blob/master/src/components/shared/Accordion.js#L60 I would put a breakpoint in the debugger to see why the handler does not work for when you press space/enter.

**Comment by yzen** [comment]
@nastark519 hi, let me know if there you had any more follow up questions, or if you needed more pointers.

**Comment by nastark519** [comment]
Hi there community. We are still were trying to inject a line of code above the https://github.com/devtools-html/debugger.html/blob/master/src/components/shared/Accordion.js#L60 that essentially is supposed to add an action listener for the space bar. Though we continually are getting error messages is there somewhere else that we should add an listener?

**Comment by jasonLaster** [comment]
CC @yzen @darkwing @nchevobbe

**Comment by nastark519** [comment]
We have a bug fix though the we can expand and collapse the headers with any key press not only the space bar. would this still count. No functionality has been compromised. We can still change selection with `tab` so is this good?

**Comment by darkwing** [comment]
Sure, feel free to open a pull request and we can start the discussion there! 👍

**Comment by nastark519** [comment]
Thank you @darkwing pull request #6389 was made. Will see what happens from here.
